numpy/core/tests/test_ufunc.py

```diff
@@ -610,7 +610,17 @@ def test_broadcast(self):
             warnings.simplefilter("always")
             u += v
             assert_equal(len(w), 1)
-            assert_(x[0,0]  != u[0, 0])
+            assert_(x[0, 0] != u[0, 0])
+
+        # Output reduction should not be allowed.
+        # See gh-15139
+        a = np.arange(6).reshape(3, 2)
+        b = np.ones(2)
+        out = np.empty(())
+        assert_raises(ValueError, umt.inner1d, a, b, out)
+        out2 = np.empty(3)
+        c = umt.inner1d(a, b, out2)
+        assert_(c is out2)
```

Comment:
are all these tests supposed to work today. I am asking these because I tried and atleast a subset of them seem to be working on the master. I was looking for a test that breaks today, that was fixed by this PR.

Reply:
Hmmm, the error should not be raised on master? The second one is the correct version, hmm, I don't remember, it could be that it is well tested above but I guess there were few `out` tests?

Reply:
The nditer tests just excercise the code branch where I did some maintenance here (I am happy to roll those changes out if it helps get the rest in.)

Reply:
I mean  I tried the ones in test_reduce_wrong_dimension_output also and that passed too . which test should i try to see what fails today and fixed by this PR.

Reply:
```
from numpy.core import _umath_tests as umt
a = np.arange(6).reshape(3, 2)
b = np.ones(2)
out = np.empty(())
umt.inner1d(a, b, out)
```
does not raise an error?

Reply:
ya this doesnt. my version is 1.20.0.dev0+ba4e4b6 , also tried on 1.18.1

Reply:
woops, sorry, that is supposed to raise an error :). But, I am not immediatley sure about `out=np.empty((3,3))` or so (out being too large). I guess that should raise as well or broadcast the inputs to the output (which half makes sense), may be good to add a test for it either way.

Reply:
These tests come from my PR at #15142 and were partially meant to increase the test coverage, i.e., include things that were already OK but not tested anywhere.

Reply:
@seberg - I think the test is fine as is: a is 2-dimensional so the expected output shape is (3,)

Reply:
> woops, sorry, that is supposed to raise an error :). But, I am not immediatley sure about `out=np.empty((3,3))` or so (out being too large). I guess that should raise as well or broadcast the inputs to the output (which half makes sense), may be good to add a test for it either way.

i am not sure what the correct behavior is, but this doesnt seem to raise an error with this change or broadcast the inputs to the output.

```
import numpy as np
from numpy.core import _umath_tests as umt
from numpy.testing import assert_raises
a = np.arange(6).reshape(3, 2)
b = np.ones(2)
out = np.empty((5, 3))
umt.inner1d(a, b, out)
print(out)
```
Result:
```
[[ 1.00000000e+000  5.00000000e+000  9.00000000e+000]
 [ 6.91217735e-310 -1.45681599e+144 -1.45681599e+144]
 [-1.45681599e+144 -1.45681599e+144 -1.45681599e+144]
 [-1.45681599e+144 -1.45681599e+144 -1.45681599e+144]
 [-1.45681599e+144 -1.45681599e+144  6.32404027e-322]]
```
First row is the output, rest is garbage.

Reply:
Hmm, that is not brilliant, indeed, rather seems like a bug to me! Checking with older versions (specifically, 1.17.4), though, I see that those had the same behaviour, so my sense would be to make this a separate issue, as this PR is rather large as it is already.

Reply:
sure, a seperate pr sounds good to me @mhvk .

Reply:
@anirudh2290 - would you be willing to raise an issue so we do not forget?  Thanks!

Reply:
Agreed, its a problem. I think it may be problem we were discussing below? But there is probably more to it. This PR really only addresses the issue of non-existing axes (and generalize the approach to existing axes to also support `keepdims=True`).

To be clear, for `-1` (gufuncs) itself, the change is fairly small. In almost all cases `-1` currently does correctly mean "reduction (ok) axes". But, we also use it for unspecified output axes, which should normally be allocated using the iterator shape.
A clash of meanings which seems small for gufuncs, but was carefully worked around in reductions.

The PR resolves this, by marking all dimensions (-1 or not) if they are considered to be "reduced". And in that case, we know that the shape must be 1, even if we allocate the array in the iterator.
This is what makes the flagging approach so nice.  Not only does it mean we can push the "not too few output dimensions" check off into the iterator (which is all that this actually fixes for gufuncs). It also means the iterator can now allocate reduction results correctly even for arbitrarily complex reductions, and including `keepdims`.